src/Mvc/src/Microsoft.AspNetCore.Mvc.RazorPages/Builder/RazorPagesEndpointRouteBuilderExtensions.cs

```diff
@@ -1,42 +1,50 @@
 ﻿// Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
 using System.Linq;
-using Microsoft.AspNetCore.Mvc.RazorPages;
-using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Mvc.RazorPages.Infrastructure;
 using Microsoft.AspNetCore.Routing;
 using Microsoft.Extensions.DependencyInjection;
 
 namespace Microsoft.AspNetCore.Builder
 {
     public static class RazorPagesEndpointRouteBuilderExtensions
     {
-        public static IEndpointConventionBuilder MapRazorPages(
-            this IEndpointRouteBuilder routeBuilder,
-            string basePath = null)
+        public static IEndpointConventionBuilder MapRazorPages(this IEndpointRouteBuilder routes)
         {
-            var mvcEndpointDataSource = routeBuilder.DataSources.OfType<MvcEndpointDataSource>().FirstOrDefault();
-
-            if (mvcEndpointDataSource == null)
+            if (routes == null)
             {
-                mvcEndpointDataSource = routeBuilder.ServiceProvider.GetRequiredService<MvcEndpointDataSource>();
-                routeBuilder.DataSources.Add(mvcEndpointDataSource);
+                throw new ArgumentNullException(nameof(routes));
             }
 
-            var conventionBuilder = new DefaultEndpointConventionBuilder();
+            EnsureRazorPagesServices(routes);
+
+            return GetOrCreateDataSource(routes);
+        }
 
-            mvcEndpointDataSource.AttributeRoutingConventionResolvers.Add(actionDescriptor =>
+        private static void EnsureRazorPagesServices(IEndpointRouteBuilder routes)
+        {
+            var marker = routes.ServiceProvider.GetService<PageActionEndpointDataSource>();
+            if (marker == null)
             {
-                if (actionDescriptor is PageActionDescriptor pageActionDescriptor)
-                {
-                    // TODO: Filter pages by path
-                    return conventionBuilder;
-                }
+                throw new InvalidOperationException(Mvc.Core.Resources.FormatUnableToFindServices(
+                    nameof(IServiceCollection),
+                    "AddMvc",
+                    "ConfigureServices(...)"));
+            }
+        }
 
-                return null;
-            });
+        private static PageActionEndpointDataSource GetOrCreateDataSource(IEndpointRouteBuilder routes)
+        {
+            var dataSource = routes.DataSources.OfType<PageActionEndpointDataSource>().FirstOrDefault();
```

Comment:
Off topic for this PR, but what do you think of this pattern? It feels like some ceremony that endpoint routing requires each framework to jump through

Reply:
Yes, and I think it's fine. 99.99% of people that touch this code will be app developers, not framework authors. I don't want to put something in the users' face that they wouldn't use.